### ISSUE TITLE
Confusion regarding visual state transitions with escape key theme disabled

### ISSUE THREAD

**Issue by bpstahlman** [noctuid/lispyville#17]
I've been using `(escape insert)` in my key themes, but as I began to learn more about the non-Lispy functionality provided by Lispyville, I found myself wanting to use operators like `y` and `d` in visual state. But then it occurred to me that I was never really in evil visual state. Because I also have `(mark normal)` in my key themes, `v`, `V` and `M-m` all put me into evil insert state, and the aforementioned escape theme ensures that I transition directly back to evil normal state when I cancel the region with escape. With the escape key theme _disabled_, if I mark a symbol in insert state, a subsequent escape transitions me back to normal state (according to the `<N>` in the status line), but leaves the region active and allows me to use commands like `y` to operate on the region. Also, subsequent motions move the edge of the region as it would in evil. This would seem to indicate I'm in visual state, though the status area gives no indication of it (e.g., no `-- VISUAL --`). But what really confused me was the fact that if I activate the mark in _normal_ state (e.g., with `M-m`), the status lines look the same as just described (transition to `<I>` on mark, then back to `<N>` on escape), but the region disappears after the escape. Clearly, there's something different going on here, which isn't reflected in the modes displayed in the status area. Another thing that surprises me about this (apparent) visual state (and makes me wonder whether it's really true visual state) is that I can't get out of it with escape, but have to hit `C-g` to deactivate the region. Interestingly, hitting `C-g` instead of escape will take me from lispy region special all the way back to normal state, skipping right over the visual state normally entered on escape.

Any light you can shed on what's going on here would be greatly appreciated. I'm trying to settle on a set of key themes that best fits my workflow, and would like to understand a bit more about the actual modes and transitions in play here before making a decision. My original motivation for looking at this was the following use case: I've marked a sexp and want to operate on it somehow: e.g., perhaps I want to copy it (and don't like having to use `M-w`), or perhaps I want to change it (and don't really like having to reach for something like `M-DEL`). Being able to hit `y` or `c` in this scenario would be nice. Of course, I'm thinking I'd have to weigh this convenience against the inconvenience of having to use `C-g` instead of the more convenient `C-[` to cancel the region when I *don't* want to operate on it...

**Comment by noctuid** [comment]
> though the status area gives no indication of it

I'd recommend relying on cursor color/shape (e.g. by changing `evil-visual-state-cursor`) instead of on the echo area or modeline. The echo area gets overwritten quickly, and if you mainly rely on cursor color, you won't ever have to look at the modeline to know what state you are in.

As for what's going on here, when you hit `ESC`, you are transitioning from insert to normal state (not to visual state). This is why another `ESC` does not deactivate the region. I've updated the documentation to be clear about this.

> I've marked a sexp and want to operate on it somehow: e.g., perhaps I want to copy it (and don't like having to use M-w), or perhaps I want to change it (and don't really like having to reach for something like M-DEL).

You can still use lispy's special keys for copying and deleting (`n` and `DEL` by default; you can change the bindings if you wish). I've also just added a `mark-toggle` key theme to allow toggling between special and visual state.

**Comment by bpstahlman** [comment]
> I'd recommend relying on cursor color/shape (e.g. by changing evil-visual-state-cursor) instead of on the echo area or modeline. The echo area gets overwritten quickly, and if you mainly rely on cursor color, you won't ever have to look at the modeline to know what state you are in.

Ah. Ok. I just inspected the value of `evil-state` and figured out the difference between the 2 scenarios... When I start out in evil insert state, `M-m` activates the lispy region but leaves me in evil insert state. But when I start out in evil normal state, `M-m` puts me in evil visual state (though `<I>` is shown in the status area). What surprised me here was that in both cases (evil insert and evil visual) the region behaved like a lispy special region: i.e., commands like `j` and `k` move the region by sexp, not in the Vim way. I guess I was thinking of evil visual and lispy region special as sort of orthogonal/mutally-exclusive, but that's clearly not the case here.

> You can still use lispy's special keys for copying and deleting (n and DEL by default; you can change the bindings if you wish). I've also just added a mark-toggle key theme to allow toggling between special and visual state.

Sounds great. I'll install and play around with it... Incidentally, I was going to make a feature suggestion regarding commands analogous to `A` and `I` for appending/inserting elements to a list, but I happened to see another thread today on a "vim-sexp-mappings-for-regular-people" key theme, which sounded like just what I was looking for. Is this something that's in active development?

**Comment by noctuid** [comment]
Let me know if there's anything else you'd like me to change, and I'll reopen this issue.

> Is this something that's in active development?

No, I haven't started on it yet. I'm currently working on another package (targets.el) in my free time, but once I'm done with that, I can work on that issue.

**Comment by bpstahlman** [comment]
Sounds good. If it's not too much trouble, would you be able to post the lispyville segment of your config at some point, just so I can do a sanity-check on my own? I'm seeing some unexpected stuff with the mark toggle feature (e.g., region "drifting" rightward - i.e., pulling in the space to the right of the symbol when I toggle from lispy special to visual), but I didn't want to post any issues if it's something stupid I've done in my config. If you prefer, you can send it to me privately...

**Comment by noctuid** [comment]
That is a legitimate issue that I forgot to add a fix for. The issue is that evil considers the character on/after the point as part of the region while emacs doesn't. It should be fixed now, but there is another issue that comes up after repeated toggles that I still need to look into. I'll fix that and add tests when I get the time.

I can still post my config  if you would like (it is very short and simple).

**Comment by bpstahlman** [comment]
Sounds good. Yes. I'd love to see your config if it's not too much trouble.
I've been changing lots of stuff around in mine lately, and just noticed
this morning that I had a 'lispyville-enter-insert-when-marking in there
that wasn't even defined in the latest version. Just want to be sure I have
a sane configuration to use as a baseline for my experimentation. I
appreciate your patience, and I apologize for importuning you so much
lately. Will do some investigation offline to get a better handle on the
various mode/state interactions before posting further...

**Comment by noctuid** [comment]
Actually I forgot that I need to depend on `evil-move-cursor-back` when determining what to do here. It seems you've set it to nil, so the newest change should work for you, but I'll also fix this when I get the chance.

**Comment by noctuid** [comment]
The aforementioned issues should all be fixed now. Let me know if you encounter any more problems.

> and just noticed this morning that I had a 'lispyville-enter-insert-when-marking in there that wasn't even defined in the latest version

Sorry about that; I made several breaking changes yesterday. Normally I'd try to not do this or declare things obsolete. I'm still experimenting with things for this package, but there probably won't be any changes to existing function/variable names in the future, but if there are, I'll make sure the old ones continue to work.

I'm considering binding better keys to get into special at some point, but this is all that I have at the moment for lispyville:
```
(use-package lispyville
  :load-path "~/src/emacs/lispyville/"
  :commands lispyville-mode
  :diminish " 🍙"
  :init
  (add-hook 'lispy-mode-hook #'lispyville-mode)
  :config
  (lispyville-set-key-theme '(operators slurp/barf-cp mark-toggle))
  (lispy-define-key lispy-mode-map "v" #'lispyville-toggle-mark-type)
  (general-evil-define-key 'normal lispyville-mode-map
    "c" (general-key-dispatch #'lispyville-change
          "o" #'noct:option-toggle/body
          "x" #'evil-exchange
          "c" (general-simulate-keys (#'lispyville-change "c")))
    ")" #'lispyville-right
    "(" #'lispyville-left))

(setq lispyville-motions-put-into-special t)
```

(I only have the `:load-path` and `:commands` because I'm using my local copy and not melpa)

**Comment by bpstahlman** [comment]
Great! Thanks. I've already started playing around with it. So far, the stuff I've tried works as expected; at least, it seems to have fixed the issues I had earlier today. One quick question (whose answer might clear some stuff up for me): who sets `evil-state` to visual when `v` or `M-m` is used in normal state? I ask because I've noticed that these commands map to the same function (`lispy-mark-symbol`) in both insert and normal states, yet visual mode is entered only when `lispy-mark-symbol` is triggered from normal state. I traced through `lispy-mark-symbol` in edebug and didn't see where the change of state (normal->visual) was happening. Is it a Lispyville hook that runs after `lispy-mark-symbol` and treats the insert/normal states differently? IIUC, Lispy doesn't know anything about evil, so it wouldn't be distinguishing between insert and normal states...
